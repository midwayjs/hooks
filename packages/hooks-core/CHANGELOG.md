# @midwayjs/hooks-core

## 3.0.0-rc.3

### Patch Changes

- Set correct version for examples

## 3.0.0-rc.2

### Patch Changes

- Build after publish

## 3.0.0-rc.1

### Patch Changes

- 71a3115: Release 3.0.0 rc version
