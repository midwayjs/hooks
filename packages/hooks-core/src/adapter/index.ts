import { createDebug, validateFunction } from '../common'
import { ResponseMetaData } from '../api/operator/http'

const debug = createDebug('hooks-core: adapter')

type FrameworkAdapter = {
  name: string
  handleResponseMetaData(metadata: ResponseMetaData[]): Promise<any>
}

export let framework: FrameworkAdapter = null

export function useFrameworkAdapter(adapter: FrameworkAdapter) {
  if (framework) {
    throw new Error(
      `FrameworkAdapter is already set, current: ${framework.name}, new: ${adapter.name}`
    )
  }
  validateFunction(
    adapter.handleResponseMetaData,
    'adapter.handleResponseMetaData'
  )
  debug('use framework adapter: %s', adapter.name)
  framework = adapter
}
