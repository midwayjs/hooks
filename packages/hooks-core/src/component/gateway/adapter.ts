import {
  IMidwayApplication,
  IMidwayContainer,
  IMidwayContext,
} from '@midwayjs/core'
import { ApiFunction, ServerRoute } from '../..'
import { ComponentConfig } from './interface'

export interface CreateApiParam {
  fn: ApiFunction
  file: string
  isExportDefault: boolean
}

export interface HooksGatewayAdapter {
  config: ComponentConfig
  container: IMidwayContainer
  app: IMidwayApplication<IMidwayContext>

  createApi(config: CreateApiParam): void
  is(route: ServerRoute): boolean
  afterCreate?(): void

  onError(ctx: any, error: any): void

  getGlobalMiddleware?(): any[]
}
