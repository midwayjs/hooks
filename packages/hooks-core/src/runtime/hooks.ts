import type { IMidwayContainer } from '@midwayjs/core'

import { useContext } from './index'

export function useLogger() {
  const ctx = useContext()
  return ctx.logger
}

export function usePlugin(key: any): any {
  const ctx = useContext()
  return ctx.app[key] || ctx[key]
}

export function useInject<T = any>(identifier: new () => T): Promise<T>
export function useInject<T = any>(identifier: string): Promise<any>
export async function useInject(identifier) {
  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getAsync(identifier)
}

export function useConfig(key?: string) {
  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getConfigService().getConfiguration(key)
}
