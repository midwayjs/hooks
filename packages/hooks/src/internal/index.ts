export * from '@midwayjs/hooks-internal'
