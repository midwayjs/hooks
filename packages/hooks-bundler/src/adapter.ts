import { AbstractRouter, ApiRoute, HttpTriggerType } from '@midwayjs/hooks-core'
import { AbstractBundlerAdapter, requireWithoutCache } from '@midwayjs/bundler'
import { join } from 'upath'
import { createExpressDevPack } from '@midwayjs/dev-pack'
import {
  getRouter,
  getSource,
  normalizeUrl,
  ProjectConfig,
} from '@midwayjs/hooks/internal'
import { getConfig, getProjectRoot } from '@midwayjs/hooks-config'
import type { ViteDevServer } from 'vite'
import cloneDeep from 'lodash/cloneDeep'

export class MidwayBundlerAdapter extends AbstractBundlerAdapter {
  protected root: string
  protected source: string
  protected router: AbstractRouter

  protected projectConfig: ProjectConfig

  constructor() {
    const source = getSource({ useSourceFile: true })
    const router = getRouter(source)
    super({
      name: '@midwayjs/hooks-bundler',
      router,
    })
    this.root = getProjectRoot()
    this.projectConfig = getConfig()
    this.source = source
    this.router = router
  }

  override transformApiRoutes(apis: ApiRoute[]): ApiRoute[] {
    apis = cloneDeep(apis)
    for (const api of apis) {
      if (api.trigger.type === HttpTriggerType) {
        api.trigger.path = normalizeUrl(this.router, api)
      }
    }
    return apis
  }

  override getSource(): string {
    return this.source
  }

  override getUnplugin() {
    const ctx = this
    return {
      vite: {
        resolveId(_: string, importer: string) {
          if (
            process.env.NODE_ENV !== 'production' &&
            importer &&
            ctx.router.isSourceFile(importer, ctx.source) &&
            ctx.router.isApiFile({
              mod: requireWithoutCache(importer),
              file: importer,
            })
          ) {
            return 'MIDWAY_HOOKS_VIRTUAL_FILE'
          }

          return null
        },
        async configureServer(server: ViteDevServer) {
          const { middleware } = await createExpressDevPack({
            cwd: ctx.root,
            sourceDir: join(ctx.root, ctx.projectConfig.source),
            watch: true,
          })

          server.middlewares.use(middleware)
        },
        config() {
          return {
            optimizeDeps: {
              include: ['@midwayjs/rpc'],
            },
            build: {
              manifest: true,
            },
          }
        },
      },
    }
  }
}
