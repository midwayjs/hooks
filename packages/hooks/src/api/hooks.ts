import type { IMidwayContainer } from '@midwayjs/core'
import {
  MidwayConfigService,
  getCurrentAsyncContextManager,
  ASYNC_CONTEXT_KEY,
} from '@midwayjs/core'
import { ERR_INVALID_ARG_TYPE, validateString } from '@midwayjs/hooks-core'

export function useContext<T = any>(): T {
  const ctx = getCurrentAsyncContextManager()
    .active()
    .getValue(ASYNC_CONTEXT_KEY)
  return ctx as T
}

export function useLogger() {
  const ctx = useContext()
  return ctx.logger
}

export function usePlugin(key: string): any {
  validateString(key, 'key')

  const ctx = useContext()
  return ctx.app[key] || ctx[key]
}

export function useInject<T = any>(identifier: new () => T): Promise<T>
export function useInject<T = any>(identifier: string): Promise<any>
export async function useInject(identifier) {
  if (typeof identifier !== 'function' && typeof identifier !== 'string') {
    throw new ERR_INVALID_ARG_TYPE('identifier', 'class, string', identifier)
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.getAsync(identifier)
}

export function useConfig(key?: string) {
  if (key !== undefined) {
    validateString(key, 'key')
  }

  const ctx = useContext()
  const requestContext: IMidwayContainer = ctx['requestContext']
  return requestContext.get(MidwayConfigService).getConfiguration(key)
}
