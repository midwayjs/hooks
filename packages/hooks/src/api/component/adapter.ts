import {
  AbstractRouter,
  ApiRoute,
  ContextManager,
  createDebug,
  EXPORT_DEFAULT_FUNCTION_ALIAS,
  HooksMiddleware,
  HttpTrigger,
  ResponseMetaData,
  ResponseMetaType,
  validateOneOf,
} from '@midwayjs/hooks-core'
import type { IMidwayApplication, IMidwayContainer } from '@midwayjs/core'
import {
  MidwayApplicationManager,
  MidwayServerlessFunctionService,
  MidwayWebRouterService,
  All,
  Del,
  Get,
  Head,
  Options,
  Patch,
  Post,
  Put,
  ServerlessTriggerType,
} from '@midwayjs/core'
import { FileSystemRouter, normalizePath } from '@midwayjs/hooks-internal'
import camelCase from 'lodash/camelCase'
import { ServerlessTrigger } from '../operator/serverless'
import { useHooksMiddleware } from '../middleware'
import { useContext } from '../hooks'

const debug = createDebug('hooks: MidwayFrameworkAdapter')

export interface MidwayApplication extends IMidwayApplication {
  use?: (middleware: any) => void
}

export class MidwayFrameworkAdapter {
  constructor(
    public source: string,
    public router: AbstractRouter,
    public app: MidwayApplication
  ) {}

  async initialize(
    container: IMidwayContainer,
    apis: ApiRoute[],
    globalMiddlewares: HooksMiddleware[]
  ) {
    const applicationManager = await container.getAsync(
      MidwayApplicationManager
    )

    const faas = applicationManager.getApplication('faas')
    const koa = applicationManager.getApplication('koa')

    if (faas) {
      this.app = faas
      this.createServerlessApi(
        apis,
        await container.getAsync(MidwayServerlessFunctionService)
      )
    } else {
      this.app = koa
      this.createWebApi(apis, await container.getAsync(MidwayWebRouterService))
    }

    this.registerGlobalMiddleware(globalMiddlewares)
  }

  isHttpTrigger(api: ApiRoute): api is ApiRoute<HttpTrigger> {
    return api.trigger.type === 'HTTP'
  }

  isServerlessTrigger(api: ApiRoute): api is ApiRoute<ServerlessTrigger> {
    return typeof (api.trigger as ServerlessTrigger)?.parseArgs === 'function'
  }

  createWebApi(
    apis: ApiRoute[],
    midwayWebRouterService: MidwayWebRouterService
  ) {
    for (const api of apis) {
      const providerId = this.getUniqueProviderId(api)

      debug('create trigger: %s, providerId: %s', api.trigger.type, providerId)

      if (this.isHttpTrigger(api)) {
        const http = this.createHttpApi(api)
        midwayWebRouterService.addRouter(http.handler, {
          url: http.path,
          middleware: http.middleware,
          requestMethod: http.method,
        })
      }
    }
  }

  createHttpApi(api: ApiRoute<HttpTrigger>) {
    validateOneOf(
      api.trigger.method,
      'trigger.method',
      Object.keys(this.methodDecorators)
    )

    const parseHTTPArgs = ({ ctx }) => {
      return ctx.request?.body?.args || []
    }

    return {
      path: normalizePath(this.router, api),
      middleware: api.middleware?.map(useHooksMiddleware),
      handler: this.createHandler(api.fn, parseHTTPArgs),
      method: api.trigger.method.toLowerCase(),
    }
  }

  private createHandler(
    fn: Function,
    parseArgs: (inputs: { ctx: any; args: any[] }) => any[]
  ) {
    return (ctx: any, ...args: unknown[]) => {
      return fn(...parseArgs({ ctx, args }))
    }
  }

  createServerlessApi(
    apis: ApiRoute[],
    midwayServerlessFunctionService: MidwayServerlessFunctionService
  ) {
    for (const api of apis) {
      const providerId = this.getUniqueProviderId(api)

      debug('create trigger: %s, providerId: %s', api.trigger.type, providerId)

      if (this.isHttpTrigger(api)) {
        const http = this.createHttpApi(api)
        midwayServerlessFunctionService.addServerlessFunction(http.handler, {
          type: ServerlessTriggerType.HTTP,
          metadata: {
            name: ServerlessTriggerType.HTTP,
            method: http.method as any,
            path: http.path,
            middleware: http.middleware,
          },
          functionName: providerId,
          handlerName: `${providerId}.handler`,
        })
        continue
      }

      if (this.isServerlessTrigger(api)) {
        midwayServerlessFunctionService.addServerlessFunction(
          this.createHandler(api.fn, api.trigger.parseArgs),
          {
            name: api.trigger.type,
            type: api.trigger.type,
            metadata: api.trigger.options,
            functionName: providerId,
            handlerName: `${providerId}.handler`,
          } as any
        )
      }
    }
  }

  private getUniqueProviderId(api: ApiRoute) {
    if (this.router instanceof FileSystemRouter) return api.functionId
    if (api.functionName !== EXPORT_DEFAULT_FUNCTION_ALIAS)
      return camelCase(api.functionId)

    // api router & export default function
    const router = new FileSystemRouter({
      source: this.source,
      routes: [],
    })
    return router.getFunctionId(api.file, api.functionName, true)
  }

  private methodDecorators = {
    GET: Get,
    POST: Post,
    PUT: Put,
    DELETE: Del,
    PATCH: Patch,
    HEAD: Head,
    OPTIONS: Options,
    ALL: All,
  }

  registerGlobalMiddleware(middlewares: HooksMiddleware[] = []) {
    for (const mw of middlewares) {
      this.app.use?.(useHooksMiddleware(mw))
    }
  }

  private async useUniversalRuntime(ctx: any, next: any) {
    return await ContextManager.run({ ctx }, async () => await next())
  }

  async handleResponseMetaData(metadata: ResponseMetaData[]): Promise<any> {
    const ctx = useContext()

    for (const meta of metadata) {
      switch (meta.type) {
        case ResponseMetaType.CODE:
          ctx.status = meta.code
          break
        case ResponseMetaType.HEADER:
          ctx.set(meta.header.key, meta.header.value)
          break
        case ResponseMetaType.CONTENT_TYPE:
          ctx.type = meta.contentType
          break
        case ResponseMetaType.REDIRECT:
          ctx.status = meta.code || 302
          ctx.redirect(meta.url)
          break
      }
    }
  }
}
