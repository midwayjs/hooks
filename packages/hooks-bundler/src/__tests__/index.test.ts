import { HttpTriggerType } from '@midwayjs/hooks-core'
import { setConfig, setProjectRoot } from '@midwayjs/hooks/internal'
import { MidwayBundlerAdapter } from '../adapter'

describe('file system router', () => {
  beforeEach(() => {
    setProjectRoot('/')
    setConfig({
      source: '/',
      routes: [
        {
          baseDir: '/api',
          basePath: '/api',
        },
      ],
    })
  })

  test('MidwayBundlerAdapter', async () => {
    const adapter = new MidwayBundlerAdapter()
    const apis: any = ['/foo', '/bar'].map((path) => ({
      file: '/api/foo.ts',
      trigger: { type: HttpTriggerType, path },
    }))

    adapter.transformApiRoutes(apis)
    adapter.transformApiRoutes(apis)
    adapter.transformApiRoutes(apis)

    const result = adapter.transformApiRoutes(apis)
    expect(result).toMatchSnapshot()
  })
})
