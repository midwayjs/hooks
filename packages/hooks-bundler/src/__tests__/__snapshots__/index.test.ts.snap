// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`file system router MidwayBundlerAdapter 1`] = `
Array [
  Object {
    "file": "/api/foo.ts",
    "trigger": Object {
      "path": "/api/foo",
      "type": "HTTP",
    },
  },
  Object {
    "file": "/api/foo.ts",
    "trigger": Object {
      "path": "/api/bar",
      "type": "HTTP",
    },
  },
]
`;
